### ISSUE TITLE
ChannelList Recovery doesn't work properly because It couldn't process query with extra data field in ChannelDTO for local storage(CoreData)

### ISSUE THREAD

**Issue by tokirin** [GetStream/stream-chat-swift#1854]
## What did you do?
We were waiting for merging CIS-1275, and as seen it merged, tested it.
and we found bug for our usecase.

```Swift
    let queryA = ChannelListQuery(
      filter: .and([.containMembers(userIds: [userId]), .notEqual(.init(rawValue: "owner_id"), to: userId)])
    )
   let controllerA = chatClient.channeListController(query: queryA)

    let queryB = ChannelListQuery(
      filter: .and([.containMembers(userIds: [userId]), .equal(.init(rawValue: "owner_id"), to: userId)])
    )
    let controllerB = chatClient.channeListController(query: queryB)
```

We use 2 `ChatChannelListController`s with different `ChannelListQuery`.
The two `ChannelListQuery`'s are mutually exclusive as you can see.
and the queries have extra data field named `owner_id` for identifying whether the channel is created by current user or not

## What did you expect to happen?
It should synchronize correctly for both of `ChatChannelListController` when connection is recovered

## What happened instead?
When background recovery begins, two of `ChatChanneListController` begin to remove each other's `ChatChannel`s and  messages

## The Problem I found in ChannelListUpdater.swift
I started to analyze problem from `SyncRepository.swift` and found something weird
```Swift
//ChannelListUpdater.swift
guard let queryDTO = session.channelListQuery(filterHash: updatedQuery.filter.filterHash) else { return }

let localQueryCIDs = Set(queryDTO.channels.compactMap { try? ChannelId(cid: $0.cid) })
let remoteQueryCIDs = Set(channelListPayload.channels.map(\.channel.cid))
let queryAlreadySynched = remoteQueryCIDs.intersection(synchedChannelIds)

// We are going to unlink & clear those channels that are not present in the remote query,
// and that have not been synched nor watched. Those are outdated.
let cidsToRemove = localQueryCIDs
    .subtracting(queryAlreadySynched)
    .subtracting(watchedChannelIds)
```
The problem is that **the `localQueryCIDs` are same for both of above two queries.**
so I checked `ChannelDTO`'s CoreData Model(in `StreamChatModel.xcdatamodel`) and confirmed the `extraData` field is just binary field.

**So my guess is:**
the sample filter's string literal (by printing `filterHash` method) are
```Swift
//Participating Chatroom List Filter Hash 
(members IN ["currentUserId"]) AND (owner_id != "currentUserId")
//Owned Chatroom List Filter Hash
(members IN ["currentUserId"]) AND (owner_id == "currentUserId")
```
But CoreData couldn't recognize query filter about `owner_id` so It just ignores that for both of query and processes into `(members IN ["currentUserId"])`
so the results of processing two queries are same and it produces all of `ChannelID`s I participated.
but `remoteQueryCIDs` is correct, so `cidsToRemove` becomes all of local cached channels except`remoteQueryCIDs` for each of `ChatChannelListController` recovery sequence

To tell conclusion, The cause may be that current CoreData couldn't process filter using extra data in `ChannelListQuery`.
Please check for this issue.

Sincerely

## GetStream Environment
**GetStream Chat version:** CIS-1275
**GetStream Chat frameworks:** StreamChat
**iOS version:** 15.3.1
**Swift version:** 5.x.x
**Xcode version:** 13.2.1
**Device:** iPhone 12 Pro

**Comment by tokirin** [comment]
@polqf I'm sorry for reporting it before release but I have to tell you that It has problem too with just one ChatChannelListController using extra data field in ChannelListQuery.

**Comment by tokirin** [comment]
I found another bug while testing 🥲
If I had loaded channel list more than 1 page, but the refetch operation refetches just first page and the rest of channel's messages are all deleted, but are still displayed in channel list with empty messages.

**Comment by polqf** [comment]
Hi @tokirin, thank you for all your reports. Let me take a look at them and come back to you 🙏

**Comment by polqf** [comment]
Hi @tokirin,

We've identified the issue where some channels were mistakenly appearing in the other channel list and we're actively working on a fix. For now, there's something you can try:

Override ChatChannelListVC's following methods:
```
    open func controller(_ controller: ChatChannelListController, shouldAddNewChannelToList channel: ChatChannel) -> Bool {
        channel.membership != nil // This is the default implementation
    }
    
    open func controller(_ controller: ChatChannelListController, shouldListUpdatedChannel channel: ChatChannel) -> Bool {
        channel.membership != nil // This is the default implementation
    }
```

And in those, add checks to ensure the channel that is about to be added/updated contains the expected "owner_id" value.
Bear in mind this check should be as quick as possible, and you should not be performing any expensive operation here.

Let me know if this improves a bit the situation.
We'll keep you posted!

**Comment by tokirin** [comment]
Hi @polqf
Thanks for your reply, but we still have problem on that issue.
First of all, We're not using ChatChannelListVC, only ChatChannelListController.
Second, We couldn't use that delegate methods to filter channels because of pagination problem.
If we could, we've already used only 1 ChatChannelListController.
In our usecase, we have to paginate seperately whole channel list which is owned by current user and those which current user is only participating as member.

**Comment by bielikb** [comment]
Hi @tokirin ,

we just released [new version of our SDK 4.14.0](https://github.com/GetStream/stream-chat-swift/releases/tag/4.14.0) that contains the fix for your reported issue.

Feel free to check it out right now!

Thanks again for your help! 🙌

Best,
Boris
